pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+    return obj
+
+
+def read_json(fn):
+    with open(fn) as f:
+        return json.load(f, object_hook=_operator_object_hook)
+
+
+def _local_pauli_eig_prep(op: str, idx: int):
+    """
+    Generate gate sequence to prepare a the +1 eigenstate of a Pauli operator, assuming
+    we are starting from the ground state (the +1 eigenstate of Z^{\\otimes n}).
+
+    :param op: A string representation of the Pauli operator whose eigenstate we'd like to prepare.
+    :param idx: The index of the qubit that the preparation is acting on
+    :return: A program which will prepare the requested state.
+    """
+    if op == 'X':
+        gate = RY(pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(-pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+
+    return Program(gate)
+
+
+def _local_pauli_eig_meas(op, idx):
+    """
+    Generate gate sequence to measure in the eigenbasis of a Pauli operator, assuming
+    we are only able to measure in the Z eigenbasis.
+
+    """
+    if op == 'X':
+        gate = RY(-pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+    return Program(gate)
+
+
+def _ops_belong_to_a_tpb(op_code1: str, op_code2: str):
+    """
+    Given two op strings (I, X, Y, or Z) return whether they share a tensor product basis.
+
+    I.e. are they the same or is one of them 'I'.
+    """
+    if op_code1 == op_code2:
+        # Same op
+        return True
+    elif op_code1 == 'I' or op_code2 == 'I':
+        # I commutes with everything
+        return True
+    else:
+        # Otherwise, they do not commute.
+        return False
+
+
+def _all_qubits_belong_to_a_tpb(op1: PauliTerm, op2: PauliTerm):
+    """
+    Compare all qubits between two PauliTerms to see if they share an overall tensor product basis.
+    """
+    all_qubits = set(op1.get_qubits()) | set(op2.get_qubits())
+    return all(_ops_belong_to_a_tpb(op1[q], op2[q]) for q in all_qubits)
+
+
+def construct_tpb_graph(experiments: ExperimentSuite):
+    """
+    Construct a graph where an edge signifies two experiments share a TPB.
+    """
+    g = nx.Graph()
+    for exp in experiments:
+        assert len(exp) == 1, 'already grouped?'
+        exp = exp[0]
+
+        if exp not in g:
+            g.add_node(exp, count=1)
+        else:
+            g.nodes[exp]['count'] += 1
+
+    for exp1, exp2 in itertools.combinations(experiments, r=2):
+        exp1 = exp1[0]
+        exp2 = exp2[0]
+
+        if exp1 == exp2:
+            continue
+
+        if (_all_qubits_belong_to_a_tpb(exp1.in_operator, exp2.in_operator)
+                and _all_qubits_belong_to_a_tpb(exp1.out_operator, exp2.out_operator)):
+            g.add_edge(exp1, exp2)
+
+    return g
+
+
+def group_experiments(experiments: ExperimentSuite) -> ExperimentSuite:
```

Comment:
There is a subtle point about the `clique_removal(graph)` function used in this function that is not entirely captured in the doc-strings. The list of maximal cliques returned by `clique_removal(..)` will only package together disjoint sets of vertices into maximal cliques, e.g. in a “straight line” graph like 1 - 2 - 3 - 4, it will identify this list as [{1, 2}, {3, 4}], excluding {2, 3}. This might not be OK for certain other problems, but is probably fine here since we're really trying to group together commuting observables, and since commutation is a non-transitive operation (i.e. [A, B] = [B, C] = 0 does not necessarily imply [A, C] = 0), the list of maximal cliques respects this structure. In our example of the 1 - 2 - 3 - 4 graph, even though 1 and 2 commute (share an edge), and 2 and 3 commute, it's ok to group together 1 and 2, but we can't put 1 and 3 in the same group, just because 2 and 3 commute.

Reply:
Another, much less important, thing to note about the `clique_removal(..)` function is that it only returns (as the first element of the 2-tuple it returns) a single max independent set, even though there could be more than one in the graph. We don't use the max independent set here, so this is not an issue. But it's something to keep in mind if we ever do extend `clique_removal(..)`'s use for that purpose.

Reply:
the above is correct, but as you note, this is all perfectly fine for this application.